File: core-patch/src/main/scala/eu/ostrzyciel/jelly/core/patch/internal/PatchDecoderImpl.scala
package eu.ostrzyciel.jelly.core.patch.internal

import eu.ostrzyciel.jelly.core.*
import eu.ostrzyciel.jelly.core.JellyExceptions.RdfProtoDeserializationError
import eu.ostrzyciel.jelly.core.internal.*
import eu.ostrzyciel.jelly.core.patch.*
import eu.ostrzyciel.jelly.core.patch.handler.*
import eu.ostrzyciel.jelly.core.proto.v1.*
import eu.ostrzyciel.jelly.core.proto.v1.patch.*

import scala.annotation.{experimental, switch}
import scala.reflect.ClassTag

@experimental
sealed abstract class PatchDecoderImpl[TNode >: Null, TDatatype : ClassTag](
  protected override val converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
  handler: PatchHandler[TNode],
  supportedOptions: RdfPatchOptions,
) extends PatchDecoder, ProtoDecoderBase[TNode, TDatatype, Any, Any]:

  private var streamOpt: Option[RdfPatchOptions] = None
  private var isFrameStreamType: Boolean = false
  private var isPunctuatedStreamType: Boolean = false

  protected final override def getNameTableSize: Int =
    streamOpt.map(_.maxNameTableSize) getOrElse JellyOptions.smallNameTableSize
  protected final override def getPrefixTableSize: Int =
    streamOpt.map(_.maxPrefixTableSize) getOrElse JellyOptions.smallPrefixTableSize
  protected final override def getDatatypeTableSize: Int =
    streamOpt.map(_.maxDatatypeTableSize) getOrElse 20

  override final def getPatchOpt: Option[RdfPatchOptions] = streamOpt

  private final def setPatchOpt(opt: RdfPatchOptions): Unit =
    if streamOpt.isEmpty then
      streamOpt = Some(opt)
      isFrameStreamType = opt.streamType.isFrame
      isPunctuatedStreamType = opt.streamType.isPunctuated

  override final def ingestFrame(frame: RdfPatchFrame): Unit =
    for row <- frame.rows do
      ingestRow(row)
    if isFrameStreamType then
      handler.punctuation()

  override final def ingestRow(row: RdfPatchRow): Unit =
    val r = row.row
    (row.rowType: @switch) match
      case RdfPatchRow.OPTIONS_FIELD_NUMBER => handleOptions(r.asInstanceOf[RdfPatchOptions])
      case RdfPatchRow.STATEMENT_ADD_FIELD_NUMBER => handleStatementAdd(r.quad)
      case RdfPatchRow.STATEMENT_DELETE_FIELD_NUMBER => handleStatementDelete(r.quad)
      case RdfPatchRow.NAMESPACE_ADD_FIELD_NUMBER =>
        val nsRow = r.asInstanceOf[RdfPatchNamespace]
        handler.addNamespace(
          nsRow.nsName,
          nameDecoder.decode(nsRow.value),
          decodeNsIri(nsRow.graph),
        )
      case RdfPatchRow.NAMESPACE_DELETE_FIELD_NUMBER =>
        val nsRow = r.asInstanceOf[RdfPatchNamespace]
        handler.deleteNamespace(
          nsRow.nsName,
          decodeNsIri(nsRow.value),
          decodeNsIri(nsRow.graph),
        )
      case RdfPatchRow.TRANSACTION_START_FIELD_NUMBER => handler.transactionStart()
      case RdfPatchRow.TRANSACTION_COMMIT_FIELD_NUMBER => handler.transactionCommit()
      case RdfPatchRow.TRANSACTION_ABORT_FIELD_NUMBER => handler.transactionAbort()
      case RdfPatchRow.NAME_FIELD_NUMBER => nameDecoder.updateNames(r.name)
      case RdfPatchRow.PREFIX_FIELD_NUMBER => nameDecoder.updatePrefixes(r.prefix)
      case RdfPatchRow.DATATYPE_FIELD_NUMBER =>
        val dtRow = r.datatype
        dtLookup.update(dtRow.id, converter.makeDatatype(dtRow.value))
      case RdfPatchRow.HEADER_FIELD_NUMBER =>
        val hRow = r.asInstanceOf[RdfPatchHeader]
        // ! No support for repeated terms in the header
        handler.header(hRow.key, convertTerm(hRow.value))
      case RdfPatchRow.PUNCTUATION_FIELD_NUMBER =>
        if this.isPunctuatedStreamType then handler.punctuation()
        else throw new RdfProtoDeserializationError(
          "Unexpected punctuation row in non-punctuated stream."
        )
      case _ =>
        throw new RdfProtoDeserializationError("Row kind is not set or unknown: " + row.rowType)

  private def decodeNsIri(iri: RdfIri): TNode =
    if iri == null then null
    else nameDecoder.decode(iri)

  protected def handleOptions(opt: RdfPatchOptions): Unit =
    JellyPatchOptions.checkCompatibility(opt, supportedOptions)
    setPatchOpt(opt)

  protected def handleStatementAdd(statement: RdfQuad): Unit

  protected def handleStatementDelete(statement: RdfQuad): Unit


@experimental
object PatchDecoderImpl:
  private[core] final class TriplesDecoder[TNode >: Null, TDatatype : ClassTag](
    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
    handler: TriplePatchHandler[TNode],
    supportedOptions: RdfPatchOptions,
  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):

    override protected def handleOptions(opt: RdfPatchOptions): Unit =
      if !opt.statementType.isTriples then
        throw new RdfProtoDeserializationError(
          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
          f"decoder. Only TRIPLES streams are accepted.")
      super.handleOptions(opt)

    override protected def handleStatementAdd(statement: RdfQuad): Unit =
      handler.addTriple(
        convertTermWrapped(statement.subject, lastSubject),
        convertTermWrapped(statement.predicate, lastPredicate),
        convertTermWrapped(statement.`object`, lastObject),
      )

    override protected def handleStatementDelete(statement: RdfQuad): Unit =
      handler.deleteTriple(
        convertTermWrapped(statement.subject, lastSubject),
        convertTermWrapped(statement.predicate, lastPredicate),
        convertTermWrapped(statement.`object`, lastObject),
      )

  private[core] final class QuadsDecoder[TNode >: Null, TDatatype : ClassTag](
    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
    handler: QuadPatchHandler[TNode],
    supportedOptions: RdfPatchOptions,
  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):

    override protected def handleOptions(opt: RdfPatchOptions): Unit =
      if !opt.statementType.isQuads then
        throw new RdfProtoDeserializationError(
          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
          f"decoder. Only QUADS streams are accepted.")
      super.handleOptions(opt)

    override protected def handleStatementAdd(statement: RdfQuad): Unit =
      handler.addQuad(
        convertTermWrapped(statement.subject, lastSubject),
        convertTermWrapped(statement.predicate, lastPredicate),
        convertTermWrapped(statement.`object`, lastObject),
        convertGraphTermWrapped(statement.graph),
      )

    override protected def handleStatementDelete(statement: RdfQuad): Unit =
      handler.deleteQuad(
        convertTermWrapped(statement.subject, lastSubject),
        convertTermWrapped(statement.predicate, lastPredicate),
        convertTermWrapped(statement.`object`, lastObject),
        convertGraphTermWrapped(statement.graph),
      )

  private[core] final class AnyStatementDecoder[TNode >: Null, TDatatype : ClassTag](
    converter: ProtoDecoderConverter[TNode, TDatatype, ?, ?],
    handler: AnyPatchHandler[TNode],
    supportedOptions: RdfPatchOptions,
  ) extends PatchDecoderImpl[TNode, TDatatype](converter, handler, supportedOptions):
    private var statementType: Int = -1

    override protected def handleOptions(opt: RdfPatchOptions): Unit =
      if opt.statementType.isUnspecified then
        throw new RdfProtoDeserializationError(
          "Incoming stream has no statement type set. Cannot decode.")
      else if opt.statementType.isUnrecognized then
        throw new RdfProtoDeserializationError(
          f"Incoming stream with statement type ${opt.statementType} cannot be decoded by this " +
          f"decoder. Only TRIPLES and QUADS streams are accepted.")
      this.statementType = opt.statementType.value
      super.handleOptions(opt)

    override def handleStatementAdd(statement: RdfQuad): Unit =
      val s = convertTermWrapped(statement.subject, lastSubject)
      val p = convertTermWrapped(statement.predicate, lastPredicate)
      val o = convertTermWrapped(statement.`object`, lastObject)
      (statementType : @switch) match
        case PatchStatementType.TRIPLES.value =>
          handler.addTriple(s, p, o)
        case PatchStatementType.QUADS.value =>
          val g = convertGraphTermWrapped(statement.graph)
          handler.addQuad(s, p, o, g)
        case _ =>
          throw new RdfProtoDeserializationError(
            f"Statement type is not set, statement add command cannot be decoded.")

    override def handleStatementDelete(statement: RdfQuad): Unit =
      val s = convertTermWrapped(statement.subject, lastSubject)
      val p = convertTermWrapped(statement.predicate, lastPredicate)
      val o = convertTermWrapped(statement.`object`, lastObject)
      (statementType : @switch) match
        case PatchStatementType.TRIPLES.value =>
          handler.deleteTriple(s, p, o)
        case PatchStatementType.QUADS.value =>
          val g = convertGraphTermWrapped(statement.graph)
          handler.deleteQuad(s, p, o, g)
        case _ =>
          throw new RdfProtoDeserializationError(
            f"Statement type is not set, statement delete command cannot be decoded.")

File: core-patch/src/main/java/eu/neverblink/jelly/core/patch/internal/PatchDecoderImpl.java
package eu.neverblink.jelly.core.patch.internal;

import static eu.neverblink.jelly.core.internal.BaseJellyOptions.*;

import eu.neverblink.jelly.core.*;
import eu.neverblink.jelly.core.internal.DecoderBase;
import eu.neverblink.jelly.core.patch.JellyPatchOptions;
import eu.neverblink.jelly.core.patch.PatchDecoder;
import eu.neverblink.jelly.core.patch.PatchHandler;
import eu.neverblink.jelly.core.proto.v1.*;
import eu.neverblink.jelly.core.proto.v1.RdfQuad;
import eu.neverblink.jelly.core.proto.v1.patch.*;

@ExperimentalApi
@InternalApi
public abstract sealed class PatchDecoderImpl<TNode, TDatatype>
    extends DecoderBase<TNode, TDatatype>
    implements PatchDecoder
{

    protected final PatchHandler<TNode> patchHandler;
    protected final RdfPatchOptions supportedOptions;

    private RdfPatchOptions currentOptions;
    private boolean isFrameStreamType = false;
    private boolean isPunctuatedStreamType = false;

    protected PatchDecoderImpl(
        ProtoDecoderConverter<TNode, TDatatype> converter,
        PatchHandler<TNode> patchHandler,
        RdfPatchOptions supportedOptions
    ) {
        super(converter);
        this.patchHandler = patchHandler;
        this.supportedOptions =
            supportedOptions != null ? supportedOptions : JellyPatchOptions.DEFAULT_SUPPORTED_OPTIONS;
    }

    @Override
    protected int getNameTableSize() {
        if (currentOptions == null) {
            return SMALL_NAME_TABLE_SIZE;
        }

        return currentOptions.getMaxNameTableSize();
    }

    @Override
    protected int getPrefixTableSize() {
        if (currentOptions == null) {
            return SMALL_PREFIX_TABLE_SIZE;
        }

        return currentOptions.getMaxPrefixTableSize();
    }

    @Override
    protected int getDatatypeTableSize() {
        if (currentOptions == null) {
            return SMALL_DT_TABLE_SIZE;
        }

        return currentOptions.getMaxDatatypeTableSize();
    }

    @Override
    public RdfPatchOptions getPatchOptions() {
        return currentOptions;
    }

    @Override
    public void ingestRow(RdfPatchRow row) {
        switch (row.getRowFieldNumber()) {
            case RdfPatchRow.OPTIONS -> handleOptions(row.getOptions());
            case RdfPatchRow.STATEMENT_ADD -> handleStatementAdd(row.getStatementAdd());
            case RdfPatchRow.STATEMENT_DELETE -> handleStatementDelete(row.getStatementDelete());
            case RdfPatchRow.NAMESPACE_ADD -> handleNamespaceAdd(row.getNamespaceAdd());
            case RdfPatchRow.NAMESPACE_DELETE -> handleNamespaceDelete(row.getNamespaceDelete());
            case RdfPatchRow.TRANSACTION_START -> handleTransactionStart();
            case RdfPatchRow.TRANSACTION_COMMIT -> handleTransactionCommit();
            case RdfPatchRow.TRANSACTION_ABORT -> handleTransactionAbort();
            case RdfPatchRow.NAME -> handleName(row.getName());
            case RdfPatchRow.PREFIX -> handlePrefix(row.getPrefix());
            case RdfPatchRow.DATATYPE -> handleDatatype(row.getDatatype());
            case RdfPatchRow.HEADER -> handleHeader(row.getHeader());
            case RdfPatchRow.PUNCTUATION -> handlePunctuation();
            default -> throw new RdfProtoDeserializationError(
                "Row kind is not set or unknown: " + row.getRowFieldNumber()
            );
        }
    }

    @Override
    public void ingestFrame(RdfPatchFrame frame) {
        for (final var row : frame.getRows()) {
            ingestRow(row);
        }

        if (isFrameStreamType) {
            patchHandler.punctuation();
        }
    }

    protected abstract void handleStatementAdd(RdfQuad statement);

    protected abstract void handleStatementDelete(RdfQuad statement);

    protected void handleOptions(RdfPatchOptions opt) {
        JellyPatchOptions.checkCompatibility(opt, supportedOptions);
        setPatchOptions(opt);
    }

    private void handleNamespaceAdd(RdfPatchNamespace nsRow) {
        final var valueIri = nsRow.getValue();
        patchHandler.addNamespace(
            nsRow.getName(),
            // The value is required for the namespace add operation
            getNameDecoder().decode(valueIri.getPrefixId(), valueIri.getNameId()),
            convertGraphTermWrapped(nsRow)
        );
    }

    private void handleNamespaceDelete(RdfPatchNamespace nsRow) {
        final var valueIri = nsRow.getValue();
        patchHandler.deleteNamespace(
            nsRow.getName(),
            // The value is not required for the namespace delete operation, null is fine
            decodeNsIri(valueIri),
            convertGraphTermWrapped(nsRow)
        );
    }

    private void handleTransactionStart() {
        patchHandler.transactionStart();
    }

    private void handleTransactionCommit() {
        patchHandler.transactionCommit();
    }

    private void handleTransactionAbort() {
        patchHandler.transactionAbort();
    }

    private void handleName(RdfNameEntry name) {
        getNameDecoder().updateNames(name);
    }

    private void handlePrefix(RdfPrefixEntry prefix) {
        getNameDecoder().updatePrefixes(prefix);
    }

    private void handleDatatype(RdfDatatypeEntry dtRow) {
        getDatatypeLookup().update(dtRow.getId(), converter.makeDatatype(dtRow.getValue()));
    }

    private void handleHeader(RdfPatchHeader hRow) {
        // No support for repeated terms in the header
        patchHandler.header(hRow.getKey(), convertTerm(hRow.getValue()));
    }

    private void handlePunctuation() {
        if (!isPunctuatedStreamType) {
            throw new RdfProtoDeserializationError("Unexpected punctuation row in non-punctuated stream.");
        }

        patchHandler.punctuation();
    }

    private void setPatchOptions(RdfPatchOptions options) {
        if (currentOptions != null) {
            return;
        }

        this.currentOptions = options;
        this.isFrameStreamType = options.getStreamType() == PatchStreamType.FRAME;
        this.isPunctuatedStreamType = options.getStreamType() == PatchStreamType.PUNCTUATED;
    }

    private TNode decodeNsIri(RdfIri iri) {
        if (iri == null) {
            return null;
        }

        return getNameDecoder().decode(iri.getPrefixId(), iri.getNameId());
    }

    @ExperimentalApi
    public static final class TriplesDecoder<TNode, TDatatype> extends PatchDecoderImpl<TNode, TDatatype> {

        private final PatchHandler.TriplePatchHandler<TNode> patchHandler;

        public TriplesDecoder(
            ProtoDecoderConverter<TNode, TDatatype> converter,
            PatchHandler.TriplePatchHandler<TNode> patchHandler,
            RdfPatchOptions supportedOptions
        ) {
            super(converter, patchHandler, supportedOptions);
            this.patchHandler = patchHandler;
        }

        @Override
        protected void handleOptions(RdfPatchOptions opt) {
            if (opt.getStatementType() != PatchStatementType.TRIPLES) {
                throw new RdfProtoDeserializationError(
                    "Incoming stream with statement type %s cannot be decoded by this decoder. Only TRIPLES streams are accepted.".formatted(
                        opt.getStatementType()
                    )
                );
            }
            super.handleOptions(opt);
        }

        @Override
        protected void handleStatementAdd(RdfQuad statement) {
            patchHandler.addTriple(
                convertSubjectTermWrapped(statement),
                convertPredicateTermWrapped(statement),
                convertObjectTermWrapped(statement)
            );
        }

        @Override
        protected void handleStatementDelete(RdfQuad statement) {
            patchHandler.deleteTriple(
                convertSubjectTermWrapped(statement),
                convertPredicateTermWrapped(statement),
                convertObjectTermWrapped(statement)
            );
        }
    }

    @ExperimentalApi
    public static final class QuadsDecoder<TNode, TDatatype> extends PatchDecoderImpl<TNode, TDatatype> {

        private final PatchHandler.QuadPatchHandler<TNode> patchHandler;

        public QuadsDecoder(
            ProtoDecoderConverter<TNode, TDatatype> converter,
            PatchHandler.QuadPatchHandler<TNode> patchHandler,
            RdfPatchOptions supportedOptions
        ) {
            super(converter, patchHandler, supportedOptions);
            this.patchHandler = patchHandler;
        }

        @Override
        protected void handleOptions(RdfPatchOptions opt) {
            if (opt.getStatementType() != PatchStatementType.QUADS) {
                throw new RdfProtoDeserializationError(
                    "Incoming stream with statement type %s cannot be decoded by this decoder. Only QUADS streams are accepted.".formatted(
                        opt.getStatementType()
                    )
                );
            }
            super.handleOptions(opt);
        }

        @Override
        protected void handleStatementAdd(RdfQuad statement) {
            patchHandler.addQuad(
                convertSubjectTermWrapped(statement),
                convertPredicateTermWrapped(statement),
                convertObjectTermWrapped(statement),
                convertGraphTermWrapped(statement)
            );
        }

        @Override
        protected void handleStatementDelete(RdfQuad statement) {
            patchHandler.deleteQuad(
                convertSubjectTermWrapped(statement),
                convertPredicateTermWrapped(statement),
                convertObjectTermWrapped(statement),
                convertGraphTermWrapped(statement)
            );
        }
    }

    @ExperimentalApi
    public static final class AnyStatementDecoder<TNode, TDatatype> extends PatchDecoderImpl<TNode, TDatatype> {

        private final PatchHandler.AnyPatchHandler<TNode> patchHandler;
        private PatchStatementType statementType = null;

        public AnyStatementDecoder(
            ProtoDecoderConverter<TNode, TDatatype> converter,
            PatchHandler.AnyPatchHandler<TNode> patchHandler,
            RdfPatchOptions supportedOptions
        ) {
            super(converter, patchHandler, supportedOptions);
            this.patchHandler = patchHandler;
        }

        @Override
        protected void handleOptions(RdfPatchOptions opt) {
            if (opt.getStatementType() == PatchStatementType.UNSPECIFIED) {
                throw new RdfProtoDeserializationError("Incoming stream has no statement type set. Cannot decode.");
            }
            if (opt.getStatementType() == null) {
                throw new RdfProtoDeserializationError(
                    "Incoming stream has an unrecognized statement type cannot be decoded by this decoder. " +
                        "Only TRIPLES and QUADS streams are accepted."
                );
            }

            this.statementType = opt.getStatementType();
            super.handleOptions(opt);
        }

        @Override
        protected void handleStatementAdd(RdfQuad statement) {
            final var s = convertSubjectTermWrapped(statement);
            final var p = convertPredicateTermWrapped(statement);
            final var o = convertObjectTermWrapped(statement);

            if (statementType == null) {
                throw new RdfProtoDeserializationError(
                    "Statement type is not set, statement add command cannot be decoded."
                );
            }
            switch (statementType) {
                case TRIPLES -> patchHandler.addTriple(s, p, o);
                case QUADS -> {
                    final var g = convertGraphTermWrapped(statement);
                    patchHandler.addQuad(s, p, o, g);
                }
            }
        }

        @Override
        protected void handleStatementDelete(RdfQuad statement) {
            final var s = convertSubjectTermWrapped(statement);
            final var p = convertPredicateTermWrapped(statement);
            final var o = convertObjectTermWrapped(statement);

            if (statementType == null) {
                throw new RdfProtoDeserializationError(
                    "Statement type is not set, statement delete command cannot be decoded."
                );
            }
            switch (statementType) {
                case TRIPLES -> patchHandler.deleteTriple(s, p, o);
                case QUADS -> {
                    final var g = convertGraphTermWrapped(statement);
                    patchHandler.deleteQuad(s, p, o, g);
                }
            }
        }
    }
}
